### ISSUE TITLE
findings: run static code analyser

### ISSUE THREAD

**Issue by mehlis** [RIOT-OS/RIOT#480]
`RIOT $ cppcheck --force .`

```
Checking boards/msba2-common/tools/src/ihex.c...
[boards/msba2-common/tools/src/ihex.c:95]: (error) Resource leak: fp

Checking dist/tools/linux-border_router/testing.c...
[dist/tools/linux-border_router/testing.c:39]: (error) Resource leak: skeleton_file

Checking sys/transceiver/transceiver.c: MODULE_AT86RF231...
[sys/transceiver/transceiver.c:336]: (error) Invalid number of character ({) when these macros are defined: 'MODULE_AT86RF231'.
```

**Comment by PeterKietzmann** [comment]
I neither tested nor had a deeper look inside bud shouldn't this issue be fixed with #2038?

**Comment by OlegHahm** [comment]
Apparently not:
```
.-(~/git/RIOT)------------------------------------------------------------------------------------(oleg@zoidberg)-
`--> ./dist/tools/cppcheck/check.sh 2>&1 |cut -d: -f3|sort|uniq -c|sort -n
      1  error (memleak)
      2  style (unusedVariable)
      5  style (variableScope)
      8  style (clarifyCalculation)
     21  error (uninitvar)
```

**Comment by kushalsingh007** [comment]
I guess all of the problems with this are solved then ? 
Because after #2665 gets merged , the only thing getting left will be (unitvariable warnings )
And to quote @gebart's  [comment] (https://github.com/RIOT-OS/RIOT/pull/2567#issuecomment-77948298)
>Just a heads up: core_cmFunc.h and some other headers in the cpu directory are supplied by ARM or the CPU vendors and may be replaced from time to time when new versions of CMSIS are released. There may be other files which are more worthwhile to spend your effort on fixing so you don't feel like you are wasting your time.

**Comment by LudwigOrtmann** [comment]
Please keep in mind that different versions of cppcheck show different things.

**Comment by kushalsingh007** [comment]
I have been working so far with ```Cppcheck 1.68``` , and my  warnings do match with the one's that have been shown above 2 comments above. Not sure if that was what you wanted to know ?

**Comment by LudwigOrtmann** [comment]
I didn't want to know anything, but remind everyone working on this issue to double-check the outcome with their respective versions of cppcheck.

**Comment by kushalsingh007** [comment]
So then I guess this one can be closed then ?

**Comment by LudwigOrtmann** [comment]
I still get `21  error (uninitvar)`.

**Comment by LudwigOrtmann** [comment]
On ubuntu 12.04 LTS:
```
      2  style (unreadVariable)
      4  style (variableScope)
     20  style (variableHidingEnum)
     57  error (nullPointer)
```

**Comment by gebart** [comment]
@LudwigOrtmann Using cppcheck 1.68 I get 21 `uninitvar`, all from the core_cmFunc.h CMSIS headers. The variables are not really uninitialized, they are set by inline asm code which cppcheck doesn't understand, and then returned. The warnings can be suppressed, but when there is a new CMSIS release the patch will have to be reapplied.

**Comment by LudwigOrtmann** [comment]
We could still exclude the CMSIS headers from being checked (not sure if this has been discussed and with what result before).

**Comment by N8Fear** [comment]
Sorry that I don't participate more actively right now, but I'm still swamped with other things... :-/
Nonetheless we already had a solution that excluded vendor files (for the unusedStructMember check a while ago). Take a look at 75c02c937a275d80ace381f9d9c13ef0ecc52020 .
This solution was IIRC abandoned because we favoured to suppress the warnings for unusedStructMembers globally. (Again IIRC) there was no technical reason involved - only the fact that the suppression file only contained unusedStructMember warnings and therefore was rendered obsolete.
The reintroduction of a suppression file suppressing errors from vendor related files is also only a trivial change to the cppcheck invocation so it should be easy to implement (see above commit).

**Comment by kushalsingh007** [comment]
@LudwigOrtmann : My version does not show those errors like
``` 20  style (variableHidingEnum) ``` 
 or ``` 57  error (nullPointer)```  is it because your version is little older ?

@N8Fear : Should I add the the entries in vendor_suppression for now ?

**Comment by LudwigOrtmann** [comment]
@kushalsingh007 yes

**Comment by LudwigOrtmann** [comment]
Interesting in this context is what version of cppcheck Travis is using. I guess it's somewhere in between.

**Comment by OlegHahm** [comment]
@zhuoshuguo, this might be a nice issue to familiarize yourself a little bit more with different parts of RIOT.

**Comment by zhuoshuguo** [comment]
@OlegHahm OK, I will try to see what is there, not quite familiar with this yet.

**Comment by kYc0o** [comment]
Is this still an issue?

**Comment by authmillenon** [comment]
Just running `./dist/tools/cppcheck/check.sh 2>&1 |cut -d: -f3|sort|uniq -c|sort -n` as defined above to see if it is.

**Comment by authmillenon** [comment]
And here are the results:

```
      1  error (bufferAccessOutOfBounds)
      1  error (syntaxError)
      1  performance (redundantCopy)
      1  style (duplicateExpressionTernary)
      1  style (knownConditionTrueFalse)
      1  style (truncLongCastAssignment)
      1  style (unsignedLessThanZero)
      1  style (unusedVariable)
      1  warning (clarifyStatement)
      1  warning (nullPointerRedundantCheck)
      1  warning (selfAssignment)
      2  style (unreachableCode)
      3  error (integerOverflow)
      3  style (unreadVariable)
      4  error (arrayIndexOutOfBounds)
      4  performance (redundantAssignment)
      4  warning (argumentSize)
      4  warning (invalidPrintfArgType_uint)
      5  warning (invalidPrintfArgType_sint)
     22  style (variableScope)
     32  error (nullPointer)
```

(Arch, `cppcheck` 1.72)

**Comment by kYc0o** [comment]
wow... they're many (much less than before)... well, I'll mark it for the next release, since several PRs were addressed to this issue and were merged, so with a few more efforts this can be done.

**Comment by authmillenon** [comment]
FIX ME FIRST is a special milestone for very old issues. It should stay at this milestone (and we should take it more serious).

**Comment by kYc0o** [comment]
Uhm... ok then, but I was more about take care of this since the milestone (in the form of a deadline, a date...) is set. So, now this means that every issue marked as FIX ME FIRST should be solved before anything?

**Comment by miri64** [comment]
Thanks to the efforts during this release we have less (but still some) findings:

```
      1  error (bufferAccessOutOfBounds)
      1  error (syntaxError)
      1  performance (redundantAssignment)
      1  performance (redundantCopy)
      1  style (duplicateExpressionTernary)
      1  style (knownConditionTrueFalse)
      1  style (truncLongCastAssignment)
      1  style (unusedVariable)
      1  warning (clarifyStatement)
      1  warning (nullPointerRedundantCheck)
      1  warning (selfAssignment)
      2  style (unreachableCode)
      2  style (unsignedLessThanZero)
      2  warning (invalidPrintfArgType_sint)
      3  error (integerOverflow)
      3  style (unreadVariable)
      3  warning (invalidPrintfArgType_uint)
      4  warning (argumentSize)
     15  style (variableScope)
     32  error (nullPointer)
```

**Comment by makomi** [comment]
I checked out the current master (d79cd42) and got these results:

    $ ./dist/tools/cppcheck/check.sh 2>&1 |cut -d: -f3|sort|uniq -c|sort -n
          1  performance (redundantCopy)
          1  style (unusedVariable)
          1  warning (clarifyStatement)
          2  style (unreachableCode)
          2  style (unsignedLessThanZero)
          2  warning (invalidPrintfArgType_uint)
          3  style (unreadVariable)
         12  style (variableScope)
         14  error (uninitvar)

Am I looking at the wrong branch?

**Comment by makomi** [comment]
Would it make sense to call `cppcheck` as part of every pull request's automatically performed checks, in order to avoid introducing ever new issues? Newly introduced warnings and errors would be handled as part of the pull request by the person that wrote the code. It would also function as a great feedback on the code's quality helping one to avoid such warnings/ errors in the future.

I figure that without such automatic checks, this issue will never really be closed.

**Comment by OlegHahm** [comment]
@makomi, that's already happening in https://github.com/RIOT-OS/RIOT/tree/master/dist/tools/cppcheck

**Comment by makomi** [comment]
@OlegHahm, it seems to me that new issues are introduced somehow, though. When I made the above pull requests the reported issues were down to this:

    $ ./dist/tools/cppcheck/check.sh 2>&1 |cut -d: -f3|sort|uniq -c|sort -n
          1  performance (redundantCopy)
          1  style (unsignedLessThanZero)
          1  style (variableScope)
          1  warning (clarifyStatement)
          2  style (unreachableCode)
          3  style (unreadVariable)

**Comment by miri64** [comment]
> Am I looking at the wrong branch?

Mine was master too, but `Cppcheck 1.72` on `Ubuntu 16.04`.

**Comment by smlng** [comment]
is this resolved with #6703 merged?

**Comment by miri64** [comment]
Yes
